Bender Book (#295)
pulp-platform/bender: book/src/workflow/dependencies.md

# Adding and Updating Dependencies

As working with dependencies is one of bender's main features, there are a few commands to ensure functionality and assist with understanding the dependency structure.

## New dependencies

> Once new dependencies are added to the manifest, bender needs to first be made aware of their existence. Otherwise, some commands will not work correctly and return an error. To update dependencies, run the following command:

Reviewer: ```suggestion
Once new dependencies are added to the manifest, bender needs to first be made aware of their existence. Otherwise, the internally used dependency tree will be incorrect or mismatched, and some commands will not work correctly, returning an error. To update dependencies, run the following command:
```

```sh
bender update
```

In case other dependencies already exist and you do not want to re-resolve these, you can add the `--new-only` flag to the update command.

> 

Reviewer: `Bender.lock` is referred to below, but not really introduced here. A quick note can fix that.

```suggestion

> **Note:** On update, bender creates or modifies the [`Bender.lock` lockfile](./lockfile.md), which keeps track of the currently selected dependency versions.

```

Reviewer: ```suggestion

> **Note:** On update, bender creates or modifies the [`Bender.lock` lockfile](./lockfile.md), which keeps track of the currently selected dependency versions.

> **Note:** Most bender commands will automatically run an update if no `Bender.lock` lockfile is found.

```

> ## Updating dependencies

Reviewer: I think somewhere in this file (or in the dependencies.md), we need to mention what update does: It picks the highest compatible semver version and unique revision. If there are conflicts in the requirements and it cannot match all criteria, it asks the user. I didn't see this in dependencies.md or workflow/dependencies.md, but I think we need it somewhere to clarify what's going on internally.

Similar to when adding new dependencies, updating existing dependencies to more recent versions is also done with the `update` command.

### Single Dependency Update
You don't always have to update your entire dependency tree. To update only a specific package, provide its name:

```sh
bender update <PKG_NAME>
```

### Recursive Updates
By default, updating a single dependency will not update its own child dependencies. If you want to update a package and all of its dependencies recursively, use the `-r/--recursive` flag:

```sh
bender update <PKG_NAME> --recursive
```

## Checking out Dependencies

> While `bender update` resolves versions and updates the lockfile, it does not necessarily download all the source code. To ensure all dependencies are locally available, use:

Reviewer: It may make sense to increase cross-linkages (e.g., lockfile here, and other spots).

Author: makes sense. I linked all `Bender.yml`, `Bender.lock` and `Bender.local` references to its section everywhere.

```sh
bender checkout
```

Bender will download the exact revisions specified in `Bender.lock`. This command is safe to run multiple times; it will only download missing packages or update those that have changed in the lockfile.

> **Note:** Many other commands (like `sources` or `script`) will automatically trigger a checkout if they detect missing dependencies.

> ## Inspecting the Dependency Tree

Reviewer: Can we link directly to the command's subfield in `commands.md`?

```suggestion
## Inspecting the Dependency Tree (`packages`)
```

To see the full list of dependencies and how they relate to each other, use the `packages` command:

```sh
bender packages
```

This prints a tree structure showing every package in your project, its resolved version, and its source. Use the `-f/--flat` flag if you just want a simple list of names and versions.

> ## Finding Package Paths

Reviewer: ```suggestion
## Finding Package Paths (`path`)
```

If you need to know where a specific dependency is stored (e.g., to point an external tool to it), use:

```sh
bender path <PKG_NAME>
```

> This will output the absolute path to the package's root directory.

Reviewer: It might make sense to reference that this is the checkout directory, and I believe the command also checks out the specific dependency if it is not already checked out.

Reviewer: done

## Checking Usage (`parents`)

If you are wondering why a specific dependency was included or why a certain version was forced, you can use:

```sh
bender parents <PKG_NAME>
```

> This will show you all packages in your tree that depend on `<PKG_NAME>` and what version constraints they have placed on it.

Reviewer: One big item missing here IMO is the `audit` command - this command can really help with checking the up-to-date-ness of dependencies as well as any potential problems.

Reviewer: done
